presto-product-tests/src/main/java/com/facebook/presto/tests/SqlCancelTests.java

```diff
@@ -0,0 +1,186 @@
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static com.facebook.presto.tests.TestGroups.CANCEL_QUERY;
+import static com.google.common.util.concurrent.Uninterruptibles.sleepUninterruptibly;
+import static com.teradata.tempto.assertions.QueryAssert.assertThat;
+import static com.teradata.tempto.query.QueryExecutor.query;
+import static java.lang.String.format;
+import static java.util.UUID.randomUUID;
+import static java.util.concurrent.Executors.newFixedThreadPool;
+import static org.testng.Assert.assertNotNull;
+import static org.testng.Assert.fail;
+
+public class SqlCancelTests
+        extends ProductTest
+{
+    private static final long WAIT_TIME = 10000;
+    private static final long STEP = 100;
+
+    private ExecutorService executor;
+    private QueryCanceller queryCanceller;
+
+    @Inject
+    @Named("databases.presto.server_address")
+    private String serverAddress;
+
+    @BeforeTestWithContext
+    public void setUp()
+            throws Exception
+    {
+        executor = newFixedThreadPool(2);
+        queryCanceller = new QueryCanceller(serverAddress);
+    }
+
+    @AfterTestWithContext
+    public void cleanUp()
+            throws Exception
+    {
+        if (executor != null) {
+            executor.shutdownNow();
+        }
+        if (queryCanceller != null) {
+            queryCanceller.close();
+        }
+    }
+
+    private class SqlExecuteCallable
+            implements Callable<Void>
+    {
+        private String sql;
+
+        public SqlExecuteCallable(String sql)
+        {
+            this.sql = sql;
+        }
+
+        public Void call()
+        {
+            assertThat(() -> query(sql)).failsWithMessage("Query was canceled");
+            return null;
+        }
+    }
+
+    private class SqlCancelCallable
+            implements Callable<Response>
+    {
+        private String sqlStatement;
+
+        public SqlCancelCallable(String sqlStatement)
+        {
+            this.sqlStatement = sqlStatement;
+        }
+
+        public Response call()
+        {
+            long endTime = System.currentTimeMillis() + WAIT_TIME;
+            while (System.currentTimeMillis() < endTime && !Thread.currentThread().isInterrupted()) {
+                QueryResult queryResult = query(format("SELECT state, query_id from system.runtime.queries" +
+                        " WHERE query = '%s' ORDER BY created DESC LIMIT 1", sqlStatement));
+                if (queryResult.getRowsCount() == 0) {
+                    sleepUninterruptibly(STEP, TimeUnit.MILLISECONDS);
+                    continue;
+                }
+                String queryState = (String) queryResult.row(0).get(0);
+                String queryId = (String) queryResult.row(0).get(1);
+                if (!queryState.equalsIgnoreCase("running")) {
```

Comment:
Why not just add `state='running'` to the select above (and remove `order by created`)?

also, then we should `LIMIT 2` and check there is at most 1 selected